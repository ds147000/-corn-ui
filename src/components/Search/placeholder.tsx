import React, { useEffect, useRef, useState } from 'react'
import { View } from '@tarojs/components'
import { transformRem } from '../../utils'

/** 不执行动画的长度 */
const STOP_ANIMTED_LEN = 1
/** 关键词展示时长 */
const ANIMTED_YIME = 5000
/** 初始化移动值 */
const DEFAULT_MOVE = 0
/** 移动速度 */
const MOVE_SEEPD = -30
/** 移动自增 */
const MOVE_ADD = 1
/** 移动最大值 */
const MOVE_MAX = 2

interface SearchPlaceholderListProps {
  data: string[]
}

const SearchPlaceholderList: React.FC<SearchPlaceholderListProps> = ({ data }) => {
  const [ move, setMove ] = useState(DEFAULT_MOVE)
  const time = useRef<NodeJS.Timeout>()

  useEffect(() => {
    if (data.length <= STOP_ANIMTED_LEN) return undefined

    function animtedMove(): void {
      time.current = setTimeout(() => {
        setMove((oldState): number => {
          if (oldState > data.length - MOVE_MAX) return DEFAULT_MOVE
          return oldState += MOVE_ADD
        })
        animtedMove()
      }, ANIMTED_YIME)
    }
    animtedMove()

    return (): void => {
      clearTimeout(time.current as NodeJS.Timeout)
    }
  }, [ data.length ])

  return (
    <View className="corn-search-placeholder" >
      <View className="corn-search-placeholder-list" style={{ top: transformRem(move * MOVE_SEEPD) }} >
        {data.map((item, key) => (
          <View key={item + String(key)} className="corn-search-placeholder-item corn-f corn-ac" >{item}</View>
        ))}
      </View>
    </View>
  )
}


export interface SearchPlaceholderProps {
  data?: string | string[]
}

export const SearchPlaceholder: React.FC<SearchPlaceholderProps> = ({ data }) => {

  if (!data) return null

  if (typeof data === 'string' || typeof data === 'number') {
    return (
      <View className="corn-search-placeholder">
        {data}
      </View>
    )
  }

  if (Array.isArray(data)) return <SearchPlaceholderList data={data} />

  return null
}

