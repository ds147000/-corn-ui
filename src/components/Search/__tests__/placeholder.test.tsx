import React, { useEffect, useState } from 'react'
import { render, waitFor } from '@testing-library/react'
import { SearchPlaceholder } from '../placeholder'
import { act } from 'react-dom/test-utils'


test('SearchPlaceholder', async () => {
  const screen = render(<SearchPlaceholder data="输入关键词" />)
  screen.getByText('输入关键词')
  await waitFor(() => expect(screen.container).toMatchSnapshot())
  screen.unmount()
})

test('SearchPlaceholder empty', async () => {
  const screen = render(<SearchPlaceholder />)
  await waitFor(() => expect(screen.container).toMatchSnapshot())
})

test('SearchPlaceholder object', async () => {
  const screen = render(<SearchPlaceholder data={{} as any} />)
  await waitFor(() => expect(screen.container).toMatchSnapshot())
})

test('SearchPlaceholder more', async () => {
  const screen = render(<SearchPlaceholder data={[
    '输入关键词',
    '小熊宝宝',
    '常青藤爸爸',
    '@Corn'
  ]} />)
  screen.getByText('输入关键词')
  screen.getByText('小熊宝宝')
  screen.getByText('常青藤爸爸')
  screen.getByText('@Corn')
  await waitFor(() => expect(screen.container).toMatchSnapshot())
})

test('SearchPlaceholder more run time', async () => {
  jest.useFakeTimers()

  const screen = render(<SearchPlaceholder data={[
    '输入关键词',
    '小熊宝宝',
    '常青藤爸爸',
    '@Corn'
  ]} />)
  screen.getByText('输入关键词')
  screen.getByText('小熊宝宝')
  screen.getByText('常青藤爸爸')
  screen.getByText('@Corn')
  act(() => {
    jest.runOnlyPendingTimers()
  })
  act(() => {
    jest.runOnlyPendingTimers()
  })
  act(() => {
    jest.runOnlyPendingTimers()
  })
  act(() => {
    jest.runOnlyPendingTimers()
  })
  act(() => {
    jest.runOnlyPendingTimers()
  })
  await waitFor(() => expect(screen.container).toMatchSnapshot())
})

test('SearchPlaceholder one', async () => {
  jest.useFakeTimers()

  const screen = render(<SearchPlaceholder data={[
    '输入关键词',
  ]} />)
  screen.getByText('输入关键词')

  screen.unmount()
})

test('SearchPlaceholder change', async () => {
  const App: React.FC = () => {
    const [list, setList] = useState(['输入关键词', '123', '456'])

    useEffect(() => {
      setList([])
    }, [])

    return <SearchPlaceholder data={list} />
  }
  const screen = render(<App />)
  await waitFor(() => expect(screen.container.querySelector('.corn-search-placeholder-item')).toBeNull())
  screen.unmount()
})


test('SearchPlaceholder more unmount', async () => {
  const screen = render(<SearchPlaceholder data={[
    '输入关键词',
    '小熊宝宝',
    '常青藤爸爸',
    '@Corn'
  ]} />)
  screen.getByText('输入关键词')
  screen.getByText('小熊宝宝')
  screen.getByText('常青藤爸爸')
  screen.getByText('@Corn')
  screen.unmount()
  await waitFor(() => expect(screen.container).toMatchSnapshot())
})


test('SearchPlaceholder unmount', async () => {
  const screen = render(<SearchPlaceholder data="输入关键词" />)
  screen.getByText('输入关键词')
  screen.unmount()
  await waitFor(() => expect(screen.container).toMatchSnapshot())
})
