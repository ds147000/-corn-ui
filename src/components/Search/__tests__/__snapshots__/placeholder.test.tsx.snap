// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPlaceholder 1`] = `
<div>
  <div
    class="corn-search-placeholder"
  >
    输入关键词
  </div>
</div>
`;

exports[`SearchPlaceholder empty 1`] = `<div />`;

exports[`SearchPlaceholder more 1`] = `
<div>
  <div
    class="corn-search-placeholder"
  >
    <div
      class="corn-search-placeholder-list"
    >
      <div
        class="corn-search-placeholder-item corn-f corn-ac"
      >
        输入关键词
      </div>
      <div
        class="corn-search-placeholder-item corn-f corn-ac"
      >
        小熊宝宝
      </div>
      <div
        class="corn-search-placeholder-item corn-f corn-ac"
      >
        常青藤爸爸
      </div>
      <div
        class="corn-search-placeholder-item corn-f corn-ac"
      >
        @Corn
      </div>
    </div>
  </div>
</div>
`;

exports[`SearchPlaceholder more run time 1`] = `
<div>
  <div
    class="corn-search-placeholder"
  >
    <div
      class="corn-search-placeholder-list"
      style="top: -0.6400rem;"
    >
      <div
        class="corn-search-placeholder-item corn-f corn-ac"
      >
        输入关键词
      </div>
      <div
        class="corn-search-placeholder-item corn-f corn-ac"
      >
        小熊宝宝
      </div>
      <div
        class="corn-search-placeholder-item corn-f corn-ac"
      >
        常青藤爸爸
      </div>
      <div
        class="corn-search-placeholder-item corn-f corn-ac"
      >
        @Corn
      </div>
    </div>
  </div>
</div>
`;

exports[`SearchPlaceholder more unmount 1`] = `<div />`;

exports[`SearchPlaceholder object 1`] = `<div />`;

exports[`SearchPlaceholder unmount 1`] = `<div />`;
